### PR TITLE
wgengine/...: split into multiple receive functions

### PR DESCRIPTION
**Pull request by josharian**
The version of wireguard-go in use here may be found at https://github.com/tailscale/wireguard-go/pull/46

Upstream wireguard-go has changed its receive model.
NewDevice now accepts a conn.Bind interface.

The conn.Bind is stateless; magicsock.Conns are stateful.
To work around this, we add a connBind type that supports
cheap teardown and bring-up, backed by a Conn.

The new conn.Bind allows us to specify a set of receive functions,
rather than having to shoehorn everything into ReceiveIPv4 and ReceiveIPv6.
This lets us plumbing DERP messages directly into wireguard-go,
instead of having to mux them via ReceiveIPv4.

One consequence of the new conn.Bind layer is that
closing the wireguard-go device is now indistinguishable
from the routine bring-up and tear-down normally experienced
by a conn.Bind. We thus have to explicitly close the magicsock.Conn
when the close the wireguard-go device.

### CODE REVIEW

File: wgengine/magicsock/magicsock.go [hunk 1]
```diff
@@ -2468,8 +2386,68 @@ func (c *Conn) DERPs() int {
...
+
+// connBind is a wireguard-go conn.Bind for a Conn.
+type connBind struct {
+	*Conn
+	mu     sync.Mutex
+	closed bool
+}
+
+// Open is called by WireGuard to create a UDP binding.
+func (c *connBind) Open(uint16) ([]conn.ReceiveFunc, uint16, error) {
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock.go:2402]: I'd name the input parameter so it looks less mysterious. Maybe named it `portWeIgnoreFromWireguard` :)

also name result parameters, at least the port.

File: wgengine/magicsock/magicsock.go [hunk 2]
```diff
@@ -2468,8 +2386,68 @@ func (c *Conn) DERPs() int {
...
+	if c.pconn6 != nil {
+		c.pconn6.SetFakeClosed(false)
+		fns = append(fns, c.receiveIPv6)
+	}
+	// TODO: Combine receiveIPv4 and receiveIPv6 and receiveIP into a single
+	// closure that closes over a *RebindingUDPConn?
+	return fns, c.LocalPort(), nil
+}
+
+func (c *connBind) SetMark(value uint32) error {
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock.go:2422]: maybe little comment about what wireguard's trying to tell us and why we do nothing?

File: wgengine/magicsock/magicsock.go [hunk 3]
```diff
@@ -2468,8 +2386,68 @@ func (c *Conn) DERPs() int {
...
+	c.closed = true
+
+	// Unblock all outstanding receives.
+	c.pconn4.SetFakeClosed(true)
+	if c.pconn6 != nil {
+		c.pconn6.SetFakeClosed(true)
+	}
+	// Send an empty read result to unblock receiveDERP,
+	// which will then check connBind.Closed.
+	c.derpRecvCh <- derpReadResult{}
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock.go:2441]: sends without a select on a closed channel always worry me. I've fought too many deadlocks from these.

Is there a closed channel we can select on? the derp reader's, the bind's, the magicsock.Conn's?

File: wgengine/magicsock/magicsock.go [hunk 4]
```diff
@@ -2468,8 +2386,68 @@ func (c *Conn) DERPs() int {
...
+	return nil
+}
+
+// Closed reports whether connBind is closed.
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock.go:2445]: "whether c" or "whether the connBind" ?

**Review by bradfitz — COMMENTED** [review]
I'm missing the purpose of the whole FakeClosed thing.

File: wgengine/magicsock/magicsock.go (hunk 3 above)
Comment (josharian) [reply wgengine/magicsock/magicsock.go:2441]: There's not, really. The wireguard-go conn.Bind compact is that it calls the receive func repeatedly until it returns an error. The only context that matters is the lifetime of the wireguard-go device. But if we select on that, then we might leak a goroutine.

Comment (josharian) [reply wgengine/magicsock/magicsock.go:2441]: But your comment did lead to a simplification and bug fix in processDERPReadResult, so thank you. :)

**Comment by josharian** [comment]
> I'm missing the purpose of the whole FakeClosed thing.

There's probably a better name, and possibly a better design. Input welcome.

wireguard-go wants binds to be stateless. It wants to be able to Close and re-Open them cheaply. And Close must cause all receive functions to immediately return an error.

magicsock conns are very stateful. The connBind is intended to be a cheap, stateless abstraction over top of the state. They must implement the Close-unblocking. The easiest way I found to do that was to have the RebindingUDPConns also support a "stateless close" or "fake close"--that is, be able to close and unblock and then re-open without actually releasing any resources.

Does that help? If not, let's discuss over video.

**Comment by bradfitz** [comment]
Thanks, that helps. Can you just drop a big comment about this somewhere in magicsock (along with a few pointers to the big comment from elsewhere if needed), as I'll surely forget all this in a month, as will future readers. Even if the comment is just echoing basically what you said here.

**Comment by josharian** [comment]
> drop a big comment

done

**Pull request merged**